app/src/mouse/pim447_trackball.c

```diff
@@ -0,0 +1,124 @@
+
+LOG_MODULE_REGISTER(PIM447, CONFIG_SENSOR_LOG_LEVEL);
+
+/*
+ * It feels more natural and more confortable to square the speed
+ * reported by the PIM447 trackball.
+ */
+static int16_t convert_speed(int32_t value)
+{
+    bool negative = false;
+
+    if (value < 0) {
+        negative = true;
+        value = -value;
+    }
+
+    if (value > 11) {
+        value = 11;
+    }
+
+    value = value * value;
+
+    if (negative) {
+        value = -value;
+    }
+
+    return value;
+}
+
+static void thread_code(void *p1, void *p2, void *p3)
+{
+    const struct device *dev;
+    int result;
+
+    /* PIM447 trackball initialization. */
+
+    const char *label = DT_LABEL(DT_INST(0, pimoroni_pim447_trackball));
+    dev = device_get_binding(label);
+    if (dev == NULL) {
+        LOG_ERR("Cannot get PIM447_TRACKBALL device");
+        return;
+    }
+
+    /* Event loop. */
+
+    bool button_press_sent   = false;
+    bool button_release_sent = false;
+
+    while (true) {
+        struct sensor_value pos_dx, pos_dy, pos_dz;
+        bool send_report = false;
+
+        result = sensor_sample_fetch(dev);
+        if (result < 0) {
+            LOG_ERR("Failed to fetch PIM447_TRACKBALL sample");
+            return;
+        }
+
+        result = sensor_channel_get(dev, SENSOR_CHAN_POS_DX, &pos_dx);
+        if (result < 0) {
+            LOG_ERR("Failed to get PIM447_TRACKBALL pos_dx channel value");
+            return;
+        }
+
+        result = sensor_channel_get(dev, SENSOR_CHAN_POS_DY, &pos_dy);
+        if (result < 0) {
+            LOG_ERR("Failed to get PIM447_TRACKBALL pos_dy channel value");
+            return;
+        }
+
+        result = sensor_channel_get(dev, SENSOR_CHAN_POS_DZ, &pos_dz);
+        if (result < 0) {
+            LOG_ERR("Failed to get PIM447_TRACKBALL pos_dz channel value");
+            return;
+        }
+
+        if (pos_dx.val1 != 0 || pos_dy.val1 != 0) {
+            zmk_hid_mouse_movement_update(convert_speed(pos_dx.val1),
+                                          convert_speed(pos_dy.val1));
+            send_report = true;
+        }
+
+        if (pos_dz.val1 == 0x80 && button_press_sent == false) {
+            zmk_hid_mouse_button_press(0);
+            button_press_sent   = true;
+            button_release_sent = false;
+            send_report = true;
+        } else if (pos_dz.val1 == 0x01 && button_release_sent == false) {
+            zmk_hid_mouse_button_release(0);
+            button_press_sent   = false;
+            button_release_sent = true;
+            send_report = true;
+        }
+
+        if (send_report) {
+            zmk_endpoints_send_mouse_report();
+            zmk_hid_mouse_clear();
+        }
+
+        k_sleep(K_MSEC(10));
+    }
+}
+
+#define STACK_SIZE 1024
+
+static K_THREAD_STACK_DEFINE(thread_stack, STACK_SIZE);
+static struct k_thread thread;
+
+int zmk_pim447_trackball_init()
+{
+    k_thread_create(&thread, thread_stack, STACK_SIZE, thread_code,
```

Comment:
Huh, this is indeed a neat trick. So that's how you run a specific function in a dedicated thread. 

I was thinking of converting the whole mouse emulation polling system to this principle instead of a workqueue, but the thread probably does not need to be constantly active? For emulation, it's trivial - we could wake the thread up only when a key is pressed, but here... Maybe we could check every two or three polling intervals, and only engage the trackball constantly if it is indeed moving?

Reply:
I don't know if a workqueue is cheaper than a thread in terms of memory or power, but it seems a bit more complex to write/maintain.

Reply:
A workqueue is the way everything is done now, and it's not suitable for mouse polling because they're not meant for time-sensitive tasks. The whole polling system needs to be rewritten to use a bare thread instead.

However, that bare thread has to work smartly. I don't think we should run the timer and the thread constantly. With emulation, we could send it to sleep when no keys are pressed and only wake it up when we have something to send (with a mutex and something like `.await()/.notify`, or however it works in Zephyr).

Here, it's more difficult to think of something, but maybe we could define a "passive state" that only polls every few intervals and wakes things up if there's any movement?